client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+    2. write(table: string, key: string, value: JSON) returns None (always succeeds)
+    3. delete(table: string, key: string) returns None
+    4. commit() returns None, throws StaleException
+- Internal Objects: 
+    1. min\_Rt: int (The minimum read time the Transaction has seen so far, based on values read)
+    2. max\_Vt: int (The maximum value time the Transaction has seen so far, based on values read)
+    3. view: list (The time-ordered list of operations performed within the Transaction to be committed to the DB server.)
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Transaction Internal Structure Description
+
+First, we must understand the purpose of the Transaction's min\_Rt (min read time) and max\_Vt (max value time).  A Transaction consists of a set of updated database entries.  The database will not allow any Transaction to commit a result derived from an inconsistent snapshot of the database values.  
+
+Specifically, every entry in a snapshot of the Treode database has both read and value timestamps.  A snapshot is consistent if and only if (maximum value timestamp) < (minimum read timestamp).  (See [this explanation](https://forum.treode.com/t/eventual-consistency-and-transactions-working-together/36).)
+
+Then, by the invariant established previously that each entry is guaranteed to be the value in the DB between [V\_t, R\_t], it follows that there was a point in time when the DB contained all the values in the snapshot simultaneously.
+
+The Transaction uses min\_Rt and max\_Vt to maintain this invariant and avoid wasting network time sending commits to the DB server guaranteed to be rejected.  If any Transaction read causes (max\_Vt > min\_Rt), the read method throws a StaleException(Vt), including the value time that caused the violation.  (Note that a read time should not cause a violation because we can simply set min-age=0. Contrarily, we cannot control the value times through our read query parameters.) 
+
+### min\_Rt 
+
+The Transaction maintains the minimum read time it has seen so far.  On every read, if the read time is less than min\_Rt, we update min\_Rt.  
+
+Note that this read time is NOT the time the Transaction read the entry.  Transaction reads from cache do not update entry read times; cache reads from the DB server do update entry read times. 
+
+### max\_Vt
+
+The Transaction maintains the maximum value time it has seen so far.  On every read, if the value time is greater than max\_Vt, we update max\_Vt.
+
+Note that updating max\_Vt can cause the invariant violation described above, causing the Transaction read method to throw StaleException(Vt).  Intuitively, when reading an entry from the DB, we have no idea when it was last updated until we actually read it.  As a result, we cannot avoid the possibility of violating our consistent snapshot invariant.
+
+* **Question (6):** If the user fails to specify max\_age correctly, should we throw a StaleException?
```

Comment:
We need to maintain the invariant max(V_t) < min(R_t).  We can't control V_t, but we can control R_t by restricting max-age.  

What if the user gives a bad max-age requirement (i.e., too loose of a requirement) such that we get a read time R_bad <= max(V_t).  Concretely, if the current time is 10, max(V_t) = 5, and the user specifies max-age = 7, then we could get value with R_bad = 4 < 5 = max(V_t). 

Is the above scenario possible or a problem?